Fix WaitForInit. Fixes #20

mberhault opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach-go/22)
<!-- Reviewable:end -->

tamird commented:
Not sure about this fix. Why does the old version work on macOS but not on Linux?

mberhault commented:
It's not a platform thing, I pulled `github.com/lib/pq` and tried on my mac, same problem.

tamird commented:
I have to rescind my earlier "this works on mac" - I just tried again and I
see the same timeout. My local lib/pq is at fcb9ef5.

On Tue, Dec 6, 2016 at 11:49 AM, marc <notifications@github.com> wrote:

> It's not a platform thing, I pulled github.com/lib/pq and tried on my
> mac, same problem.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach-go/pull/22#issuecomment-265203299>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPA72CGhGJ-mTd7e142uStd54YPd2ks5rFZIagaJpZM4LFmY5>
> .
>

mberhault commented:
hum. circleci is failing on timeout. I figured I'd drop circleci and add a teamcity build for this, but same thing.

mberhault commented:
circleci is not pulling this change, so the timeout is on the master version.
unsure about teamcity, I'm adding debugging

mberhault commented:
ok, nevermind, circleci is grabbing the change, but there's something odd:
`~/.go_workspace/src/github.com/cockroachdb/cockroach-go` (the GOPATH) is at master and is what is being used.
the actual PR code is in `~/cockroach-go/`. Even if I run manually form within `~/cockroach-go`, the test still pulls in the master version.

Is there something odd in the way we pull dependencies? This branch is the same as `examples-go`, we basically run:
```
go get -d -t ./...
go test -v -i ./...
go test -v ./...
```

Same thing in teamcity, I've added: `cat $GOPATH/..../testserver.go` and `cat testserver/testserver.go` go the build steps, and it does show differences. see https://teamcity.cockroachdb.com/viewLog.html?buildId=83306&buildTypeId=Cockroach_CockroachGo&tab=buildLog

Clearly I'm doing something wrong. @tamird, @jordanlewis, ideas?

mberhault commented:
PTAL: fixed teamcity: https://teamcity.cockroachdb.com/viewLog.html?buildId=83492&tab=buildResultsDiv&buildTypeId=Cockroach_CockroachGo
I'll drop circleCI in a subsequent PR.

tamird commented:
so...why doesn't SHOW DATABASES work?

mberhault commented:
looks like we can't talk to ourselves when using sockets:
```
I161207 14:35:23.327250 23 vendor/google.golang.org/grpc/clientconn.go:667  grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp: lookup vintner: no such host"; Reconnecting to {"vintner:50612" <nil>}
```

This used to work fine, didn't it? at the very least, the tests in `examples-go` that actually write/read data were passing

mberhault commented:
so yeah, this won't help tests actually trying to use the testserver since the whole point is to have a working cockroach node. Any idea when this may have changed?

tamird commented:
I would guess this has been broken since whenever we started trying to talk
to ourselves internally, since grpc doesn't support dialing unix sockets
without some special machinery which we do not use.

On Wed, Dec 7, 2016 at 2:45 PM, marc <notifications@github.com> wrote:

> so yeah, this won't help tests actually trying to use the testserver since
> the whole point is to have a working cockroach node. Any idea when this may
> have changed?
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach-go/pull/22#issuecomment-265552193>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPLZ86fw_Ar0zsCgeVew4q5qWE4sQks5rFwzNgaJpZM4LFmY5>
> .
>

mberhault commented:
I supposed we need to figure out if we want to try and get around that, or if we should drop sockets.
Honestly, I'm fine with doing away with them, we can use @boljen's freeport method for this. Other than that, I'm not sure we need sockets for a whole lot. @bdarnell, care to weigh in?

tamird commented:
We shouldn't be connecting to ourselves, so that's almost certainly a bug.

On Wed, Dec 7, 2016 at 2:58 PM, marc <notifications@github.com> wrote:

> I supposed we need to figure out if we want to try and get around that, or
> if we should drop sockets.
> Honestly, I'm fine with doing away with them, we can use @boljen
> <https://github.com/boljen>'s freeport method for this. Other than that,
> I'm not sure we need sockets for a whole lot. @bdarnell
> <https://github.com/bdarnell>, care to weigh in?
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach-go/pull/22#issuecomment-265555959>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPHG007PTKI8puxpSV2IIwIIP8CUrks5rFw_UgaJpZM4LFmY5>
> .
>

tamird commented:
@mberhault were you able to reproduce that connection problem outside of this package's tests? I can't see to reproduce it on current master in docker.

mberhault commented:
actually, I'm not. when I leave the same testserver that is failing, but attempt to connect to it with `cockroach sql --url='postgres://root@:26257?host=%2Ftmp%2Fcockroach-testserver331441468'`, `SHOW DATABASES` returns successfully.
does libpq persit errors somewhere?

mberhault commented:
apparently it is. if I re-open a db handle on failure, it passes on the second attempt:
```
2016/12/07 16:36:30 Opening postgres://root@:26257?host=%2Ftmp%2Fcockroach-testserver578416532
2016/12/07 16:36:30 Wait for init
2016/12/07 16:36:30 Error: pq: database "information_schema" does not exist, opening new db
2016/12/07 16:36:30 WaitForInit: <nil>
```

@mjibson: could it be that libpq is trying to use `information_schema` somewhere before the actual query and therefore fails with a persistent error? Since the process just started up, the system tables may not exist yet.

tamird commented:
Nah, lib/pq doesn't do any queries on its own - it's pretty bare. My guess
is that something gets fouled up on the cockroach side during process
startup. Perhaps it's time to bring @nvanbenschoten in.

On Wed, Dec 7, 2016 at 4:38 PM, marc <notifications@github.com> wrote:

> apparently it is. if I re-open a db handle on failure, it passes on the
> second attempt:
>
> 2016/12/07 16:36:30 Opening postgres://root@:26257?host=%2Ftmp%2Fcockroach-testserver578416532
> 2016/12/07 16:36:30 Wait for init
> 2016/12/07 16:36:30 Error: pq: database "information_schema" does not exist, opening new db
> 2016/12/07 16:36:30 WaitForInit: <nil>
>
> @mjibson <https://github.com/mjibson>: could it be that libpq is trying
> to use information_schema somewhere before the actual query and therefore
> fails with a persistent error? Since the process just started up, the
> system tables may not exist yet.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach-go/pull/22#issuecomment-265581387>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPLID0QAtrO4hGg4G-BNVI7LC3ylFks5rFydZgaJpZM4LFmY5>
> .
>

nvanbenschoten commented:
I can't seem to reproduce so it's tough to get close to the issue, but @mjibson was correct in saying that change in error text is almost certainly a result of https://github.com/cockroachdb/cockroach/pull/10196. The error is a little surprising to me, but I guess it's a result of hitting a `Server` before its SQL `Executor` has been [initialized](https://github.com/cockroachdb/cockroach/blob/83c4b3a32b11ab62703a5c078534d879d80fe6c9/pkg/sql/executor.go#L316). I'd assume we'd see the same error if we tried to query the table directly in that initialization period. Perhaps we should reconsider [the order](https://github.com/cockroachdb/cockroach/blob/master/pkg/server/server.go#L636) that we start `Server` components and perform as much setup as possible before exposing ourselves on a port.

tamird commented:
> Perhaps we should reconsider the order that we start Server components and perform as much setup as possible before exposing ourselves on a port.

Aren't we already doing that? We don't pass connections through to pgwire until:
https://github.com/cockroachdb/cockroach/blob/master/pkg/server/server.go#L647

mjibson commented on testserver/testserver.go:
```diff
@@ -225,19 +228,16 @@ func (ts *TestServer) PGURL() *url.URL {
+		log.Printf("WaitForInit: %v", err)
+		time.Sleep(time.Millisecond * 100)
 	}
+	return
```
This should return a real error since the wait didn't succeed.

mberhault replied:
it is a real error, if we broke out it means that `err != nil`

tamird commented:
I don't think we should merge this. We have a problem in Cockroach that
this papers over.

On Thu, Dec 8, 2016 at 2:22 PM, Matt Jibson <notifications@github.com>
wrote:

> *@mjibson* approved this pull request.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach-go/pull/22#pullrequestreview-12097258>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPAv-AWCGASrll8b_bip3RH93BMNvks5rGFjpgaJpZM4LFmY5>
> .
>

mberhault commented:
agreed, while the timeout doesn't hurt, using `select 1` clearly does not do what we want. The goal is to notice when the system tables are there and the cluster ready to serve, so this should go back to being `show databases`. Keeping open for now.

mberhault commented:
switched back to `SHOW DATABASES` to have a repeatable broken test to help with https://github.com/cockroachdb/cockroach/issues/12314

Also added a flag to specify a custom cockroach binary. Helpful to bisect over various binaries.

tamird commented:
LGTM, probably best to squash this and remove the "fixes #20".

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[testserver/testserver.go, line 79 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/22#-KYskxvbVAjYAdq8BniD:-KYskxvbVAjYAdq8BniE:bvhi416) ([raw file](https://github.com/cockroachdb/cockroach-go/blob/7e52d00d6f88360755bbf0fe7a4ad0d3b573b95a/testserver/testserver.go#L79)):*
> ```Go
> var (
> 	sqlURLRegexp = regexp.MustCompile("sql:\\s+(postgresql:.+)\n")
> 	customBinary string
> ```

why not `customBinary = flag.String(...)`?

---

*[testserver/testserver.go, line 147 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/22#-KYsl5p2p8QZR56v03AN:-KYsl5p2p8QZR56v03AO:b-d7fjtw) ([raw file](https://github.com/cockroachdb/cockroach-go/blob/7e52d00d6f88360755bbf0fe7a4ad0d3b573b95a/testserver/testserver.go#L147)):*
> ```Go
> 	}
> 
> 	err = ts.WaitForInit(db)
> ```

`if err := ....; err != nil {` plz

---

*[testserver/testserver.go, line 243 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/22#-KYslOc7rBCZGMgVzHAv:-KYslOc7rBCZGMgVzHAw:b-xp0bf8) ([raw file](https://github.com/cockroachdb/cockroach-go/blob/7e52d00d6f88360755bbf0fe7a4ad0d3b573b95a/testserver/testserver.go#L243)):*
> ```Go
> 
> // WaitForInit retries until a connection is successfully established.
> func (ts *TestServer) WaitForInit(db *sql.DB) (err error) {
> ```

while you're here, can you please not use a named return? I believe that's mjibson's confusion below.

---

*[testserver/testserver.go, line 244 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/22#-KYslY1v2w2EOnaayVa4:-KYslY1v2w2EOnaayVa5:bsmrfhq) ([raw file](https://github.com/cockroachdb/cockroach-go/blob/7e52d00d6f88360755bbf0fe7a4ad0d3b573b95a/testserver/testserver.go#L244)):*
> ```Go
> // WaitForInit retries until a connection is successfully established.
> func (ts *TestServer) WaitForInit(db *sql.DB) (err error) {
> 	for i := 0; i < 50; i++ {
> ```

FWIW you can use https://godoc.org/github.com/cockroachdb/cockroach/pkg/util#RetryForDuration here

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/22#-:-KYslkKt5GfIl3KqgTyd:bm7vzli)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[testserver/testserver.go, line 79 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/22#-KYskxvbVAjYAdq8BniD:-KYsmcq-wbS4jyOX4QKN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach-go/blob/7e52d00d6f88360755bbf0fe7a4ad0d3b573b95a/testserver/testserver.go#L79)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why not `customBinary = flag.String(...)`?
</blockquote></details>

Done.

---

*[testserver/testserver.go, line 147 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/22#-KYsl5p2p8QZR56v03AN:-KYsn8Yqv_wvtb8_6E3o:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach-go/blob/7e52d00d6f88360755bbf0fe7a4ad0d3b573b95a/testserver/testserver.go#L147)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

`if err := ....; err != nil {` plz
</blockquote></details>

Done.

---

*[testserver/testserver.go, line 243 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/22#-KYslOc7rBCZGMgVzHAv:-KYsnG2ukaDrOCE3_8XI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach-go/blob/7e52d00d6f88360755bbf0fe7a4ad0d3b573b95a/testserver/testserver.go#L243)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

while you're here, can you please not use a named return? I believe that's mjibson's confusion below.
</blockquote></details>

Done.

---

*[testserver/testserver.go, line 244 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/22#-KYslY1v2w2EOnaayVa4:-KYsnGahwhChcc0dIbNu:b-n8jnql) ([raw file](https://github.com/cockroachdb/cockroach-go/blob/7e52d00d6f88360755bbf0fe7a4ad0d3b573b95a/testserver/testserver.go#L244)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

FWIW you can use https://godoc.org/github.com/cockroachdb/cockroach/pkg/util#RetryForDuration here
</blockquote></details>

I can't, the whole point of this package is to avoid dependencies on `cockroachdb/cockroach`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/22)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Squashed. I'm keeping the original commit message though, I'm not merging this until this works again.

tamird commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[testserver/testserver.go, line 244 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/22#-KYslY1v2w2EOnaayVa4:-KYsntGz5v6O-o3BLF0i:bhyeweg) ([raw file](https://github.com/cockroachdb/cockroach-go/blob/7e52d00d6f88360755bbf0fe7a4ad0d3b573b95a/testserver/testserver.go#L244)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

I can't, the whole point of this package is to avoid dependencies on `cockroachdb/cockroach`
</blockquote></details>

It is? I thought the point was to not to require building cockroach.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/22)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[testserver/testserver.go, line 244 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/22#-KYslY1v2w2EOnaayVa4:-KYuDZ8wCQV4w4I91Lxp:b-pkft2b) ([raw file](https://github.com/cockroachdb/cockroach-go/blob/7e52d00d6f88360755bbf0fe7a4ad0d3b573b95a/testserver/testserver.go#L244)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

It is? I thought the point was to not to require building cockroach.
</blockquote></details>

well, both. We don't pull `cockroachdb/cockroach` in this entire repo, which keeps a nice separation. I would prefer to keep it that way, unless we have a really good reason not to. A simple retry loop doesn't seem quite like enough.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach-go/22)*
<!-- Sent from Reviewable.io -->
